clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2857,32 +2857,64 @@ static void handleWeakImportAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
 }
 
 // Checks correctness of mutual usage of different work_group_size attributes:
-// reqd_work_group_size, max_work_group_size. Values of reqd_work_group_size
-// arguments shall be equal or less than values coming from max_work_group_size.
+// reqd_work_group_size, max_work_group_size and max_global_work_dim.
+// Values of reqd_work_group_size arguments shall be equal or less than values
+// coming from max_work_group_size.
+// In case of 'max_global_work_dim' attribute equals to 0 we shall ensure
+// that if max_work_group_size and reqd_work_group_size attributes exist,
+// then they are equal (1, 1, 1).
 static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
                                      uint32_t WGSize[3]) {
+  bool Result = true;
+  if (Attr.getKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
+    if (const SYCLIntelMaxWorkGroupSizeAttr *A =
+            D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      if (A->getXDim() != 1 || A->getYDim() != 1 || A->getZDim() != 1) {
+        S.Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << Attr;
+        Result = false;
+      }
+    }
+    if (const ReqdWorkGroupSizeAttr *A =
+            D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      if (A->getXDim() != 1 || A->getYDim() != 1 || A->getZDim() != 1) {
+        S.Diag(A->getLocation(), diag::err_sycl_x_y_z_arguments_must_be_one)
+            << A << Attr;
+        Result = false;
+      }
+    }
+    return Result;
+  }
+  if (const SYCLIntelMaxGlobalWorkDimAttr *A =
+          D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+    if (A->getNumber() == 0) {
+      if (!(WGSize[0] == 1 && WGSize[1] == 1 && WGSize[2] == 1)) {
```

Comment:
There is some inconsistency with the previous analogous conditions (inverse for ANDs _vs_ earlier OR for inverses). Personally, I find this line's condition a tad more readable, so please consider unifying.

Actually, checking for this condition and emitting `err_sycl_x_y_z_arguments_must_be_one` might be factored out into a lambda of some sort, but that's _even more optional_.

Reply:
Ok, per your and @vmaksimo requests, I made a separate lambda for this check.